apps/authenticator/lib/sign_in/commands/authorization_code.ex

```diff
@@ -0,0 +1,205 @@
+defmodule Authenticator.SignIn.Commands.AuthorizationCode do
+  @moduledoc """
+  Authenticates the user identity using the Authorization Code Flow.
+
+  This flow is used in order to exchange an authorization code in an access token.
+
+  This allows clients to redirect the user to server in order to authenticate without having
+  to request users credentials on their side.
+  """
+
+  require Logger
+
+  alias Authenticator.Ports.ResourceManager, as: Port
+  alias Authenticator.Sessions
+
+  alias Authenticator.Sessions.Tokens.{
+    AccessToken,
+    AuthorizationCode,
+    ClientAssertion,
+    RefreshToken
+  }
+
+  alias Authenticator.SignIn.Commands.Inputs.AuthorizationCode, as: Input
+
+  @behaviour Authenticator.SignIn.Commands.Behaviour
+
+  @doc """
+  Sign in an user identity by Authorization Code flow.
+
+  The application has to be active, using openid-connect in order to use this flow.
+
+  When the client application has a public_key saved on database we force the use of
+  client_assertions on input to avoid passing it's secret open on requests.
+  """
+  @impl true
+  def execute(%Input{code: token, client_id: client_id, redirect_uri: redirect_uri} = input) do
+    with {:app, {:ok, app}} <- {:app, Port.get_identity(%{client_id: client_id})},
+         {:flow_enabled?, true} <- {:flow_enabled?, "authorization_code" in app.grant_flows},
+         {:app_active?, true} <- {:app_active?, app.status == "active"},
+         {:valid_protocol?, true} <- {:valid_protocol?, app.protocol == "openid-connect"},
+         {:token, {:ok, claims}} <- {:token, AuthorizationCode.verify_and_validate(token)},
+         {:same_client?, true} <- {:same_client?, client_id == claims["aud"]},
+         {:same_redirect?, true} <- {:same_redirect?, claims["redirect_uri"] == redirect_uri},
+         {:user, {:ok, user}, _} <- {:user, Port.get_identity(%{id: claims["sub"]}), claims},
+         {:user_active?, true, _} <- {:user_active?, user.status == "active", claims},
+         {:public?, true, _} <- {:public?, public_app?(app), {user, app, input, claims}} do
```

Comment:
I know this kind of implementation goes all over the code base, but recently I've been told about a [Credo recomendation ](https://hexdocs.pm/credo/Credo.Check.Readability.WithCustomTaggedTuple.html#content) to avoid it. What are your opinion about it?

Reply:
I think when you have to log out every single error in a different message it's better to add it here than  encapsulate it in functions and so on.

In my opinion it's clear enough.